src/shogun/machine/DistanceMachine.cpp

```diff
@@ -51,168 +43,69 @@ void CDistanceMachine::init()
 
 void CDistanceMachine::distances_lhs(float64_t* result,int32_t idx_a1,int32_t idx_a2,int32_t idx_b)
 {
-    // TODO: port to use OpenMP backend instead of pthread
-#ifdef HAVE_PTHREAD
-    int32_t num_threads=parallel->get_num_threads();
-#else
-    int32_t num_threads=1;
-#endif
-    ASSERT(num_threads>0)
+    int32_t num_threads;
+    int32_t num_vec;
+    int32_t step;
 
     ASSERT(result)
 
-    if (num_threads < 2)
+    #pragma omp parallel shared(num_threads, step)
     {
-        D_THREAD_PARAM param;
-        param.d=distance;
-        param.r=result;
-        param.idx_r_start=idx_a1;
-        param.idx_start=idx_a1;
-        param.idx_stop=idx_a2+1;
-        param.idx_comp=idx_b;
-
-        run_distance_thread_lhs((void*) &param);
-    }
-#ifdef HAVE_PTHREAD
-    else
-    {
-        pthread_t* threads = SG_MALLOC(pthread_t, num_threads-1);
-        D_THREAD_PARAM* params = SG_MALLOC(D_THREAD_PARAM, num_threads);
-        int32_t num_vec=idx_a2-idx_a1+1;
-        int32_t step= num_vec/num_threads;
-        int32_t t;
-
-        pthread_attr_t attr;
-        pthread_attr_init(&attr);
-        pthread_attr_setdetachstate(&attr, PTHREAD_CREATE_JOINABLE);
-
-        for (t=0; t<num_threads-1; t++)
+#ifdef HAVE_OPENMP
+        #pragma omp single
         {
-            params[t].d = distance;
-            params[t].r = result;
-            params[t].idx_r_start=t*step;
-            params[t].idx_start = (t*step)+idx_a1;
-            params[t].idx_stop = ((t+1)*step)+idx_a1;
-            params[t].idx_comp=idx_b;
-
-            pthread_create(&threads[t], &attr, CDistanceMachine::run_distance_thread_lhs, (void*)&params[t]);
+            num_threads=omp_get_num_threads();
```

Comment:
@karlnapf The variable involved need to be computed only once as they will remain constant over all the threads. In parallel the same computations will be repeated over all threads.